Builtin active record check for migrations

pcboy opened the pull request:
Hi,

I just wanted to add a new builtin check that I think is worthwhile.
We recently had rails servers running with the migrations not properly applied for some reason.
We found that `ActiveRecord::Migrator.needs_migration?` in our health check could prevent that situation.
So I added a new builtin check for that. It's separated from active_record_check.rb in order to keep retro compatibility but we can merge both if you prefer.
What do you think?

chrisarcand commented:
Seems legit to me 👍 

@anfleene?

anfleene commented:
Is that a thing in all versions of rails we support? Looks good to me otherwise.

chrisarcand commented:
Whoops, good call!

@pcboy It looks like `needs_migration?` is Active Record 4.0+
As we're currently supporting Rails 3.2, you'll need to factor that in to your changes and likely utilize [how needs_migration? is currently implemented](https://github.com/rails/rails/blob/57255d9d6a2d0f91b81a16f1a49daff458903c3e/activerecord/lib/active_record/migration.rb#L1042-L1044). 

Preferably you'd stub the test a little differently to fail on Rails 3.2 (it works only because the mock adds the missing method, hiding the failure) but if it isn't really trivial I guess I wouldn't worry too much about it.

pbyrne commented:
Is it reasonable to have the check fail (and explain why) in the absence of the method? Basically make it unhelpful in Rails 3.2 rather than explode. I wouldn't suggest a ton of effort to provide feature parity for a deprecated version of Rails.

pcboy commented:
I just added support for older ActiveRecord versions.
I added some tests but ideally that would need to be tested on an older rails version for real.

chrisarcand commented:
@pcboy Unfortunately that solution doesn't work due to underlying changes made in Rails 4:

`activerecord (3.2.22.2) lib/active_record/migration.rb:585:in 'get_all_versions'`

The above problem is easy, but I started down the path of supporting it fully and while it's totally possible it started to be overkill. You run in to edge cases of the schema migrations table not yet being created on new apps (so the check is a false positive for the FIRST migration), etc. etc. In the end I was writing half the great changes made to that Migrator interface in Rails 4.

Following my own advice (if it isn't really trivial I guess I wouldn't worry too much about it), I now think we should follow @pbyrne's suggestion and just raise or mark a failure with a good message stating that this check is simply not supported for Rails < 4.

pcboy commented:
Too bad but I was expecting this sadly. =(
No problem, I'll add the check for rails 4 this weekend.

pcboy commented:
Added  an abort directly in the initialize() so that it fails immediately if the method is not there.

pbyrne commented:
@pcboy I've submitted https://github.com/pcboy/okcomputer/pull/1 against this PR with a version I think is a bit safer. Thoughts?

pcboy commented:
@pbyrne Well, I prefer this to fail as soon as possible instead. Because your proposal will make the health check fail instead which can lead to problems, it's not exactly the same thing. With my proposal I know that it fails during the development phase, not after trying to deploy my project when the load balancer will tell me the instance is not healthy.

anfleene commented:
@pcboy I do not recommend using anything but the default check as a load balancer health check.

pcboy commented:
@anfleene Yeah maybe bad example. But we have for instance a smoke test during our deployment which is checking for /health/all as I want it to fail if the migrations are not up.
But in the end my main point is that I prefer things to fail fast, not later. If this built in check is not working on my rails version I want to know about it asap, I don't want it to work but silently fail (I'll have to check the response of the health check to know that it doesn't work properly).

pbyrne commented:
@pcboy I understand your concerns, but we're opting to prefer a failed check vs preventing the app from booting in this particular edge case.

I've merged https://github.com/sportngin/okcomputer/pull/127, which contains my tweaks to this PR and it's available as v1.14.1 of the gem. Thanks for helping to make this gem better!
